src/uu/basenc/BENCHMARKING.md

```diff
@@ -0,0 +1,177 @@
+
+To compare the runtime performance of the uutils implementation with the GNU Core Utilities implementation, you can
+use a benchmarking tool like [hyperfine][0].
+
+hyperfine currently does not measure maximum memory usage. Memory usage can be benchmarked using [poop][2], or
+[toybox][3]'s "time" subcommand (both are Linux only).
+
+Next, build the `basenc` binary using the release profile:
+
+```Shell
+cargo build --package uu_basenc --profile release
+```
+
+## Expected performance
+
+uutils' `basenc` performs streaming decoding and encoding, and therefore should perform all operations with a constant
+maximum memory usage, regardless of the size of the input. Release builds currently use less than 3 mebibytes of
+memory, and memory usage greater than 10 mebibytes should be considered a bug.
+
+As of September 2024, uutils' `basenc` has runtime performance equal to or superior to GNU Core Utilities' `basenc` in
+in most scenarios. uutils' `basenc` uses slightly more memory, but given how small these quantities are in absolute
+terms (see above), this is highly unlikely to be practically relevant to users.
+
+## Benchmark results (2024-09-27)
+
+### Setup
+
+```Shell
+# Use uutils' dd to create a 1 gibibyte in-memory file filled with random bytes (Linux only).
+# On other platforms, you can use /tmp instead of /dev/shm, but note that /tmp is not guaranteed to be in-memory.
+coreutils dd if=/dev/urandom of=/dev/shm/one-random-gibibyte bs=1024 count=1048576
+
+# Encode this file for use in decoding performance testing
+/usr/bin/basenc --base32hex -- /dev/shm/one-random-gibibyte 1>/dev/shm/one-random-gibibyte-base32hex-encoded
+/usr/bin/basenc --z85 -- /dev/shm/one-random-gibibyte 1>/dev/shm/one-random-gibibyte-z85-encoded
+```
+
+### Programs being tested
+
+uutils' `basenc`:
+
+```
+❯ git rev-list HEAD | coreutils head -n 1 -- -
+a0718ef0ffd50539a2e2bc0095c9fadcd70ab857
+```
+
+GNU Core Utilities' `basenc`:
+
+```
+❯ /usr/bin/basenc --version | coreutils head -n 1 -- -
+basenc (GNU coreutils) 9.4
+```
+
+### Encoding performance
+
+#### "--base64", default line wrapping (76 characters)
+
+➕ Faster than GNU Core Utilities
+
+```
+❯ hyperfine \
+    --sort \
+    command \
+    -- \
+    '/usr/bin/basenc --base64 -- /dev/shm/one-random-gibibyte 1>/dev/null' \
+    './target/release/basenc --base64 -- /dev/shm/one-random-gibibyte 1>/dev/null'
+
+Benchmark 1: /usr/bin/basenc --base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
+  Time (mean ± σ):     965.1 ms ±   7.9 ms    [User: 766.2 ms, System: 193.4 ms]
+  Range (min … max):   950.2 ms … 976.9 ms    10 runs
+
+Benchmark 2: ./target/release/basenc --base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
+  Time (mean ± σ):     696.6 ms ±   9.1 ms    [User: 574.9 ms, System: 117.3 ms]
+  Range (min … max):   683.1 ms … 713.5 ms    10 runs
+
+Relative speed comparison
+        1.39 ±  0.02  /usr/bin/basenc --base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
```

Comment:
I can't reproduce this:

```
Relative speed comparison
        1.00          /usr/bin/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
        1.22 ±  0.35  ./target/release/coreutils basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
        3.66 ±  1.22  ./coreutils basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
```

./coreutils is the version without the patch

Reply:
which is fine btw :)
still much better than the previous version

Reply:
"--base16" is one of the areas I found where "we" don't beat GNU. On my machine, base16 encoding is about 5-10% slower than GNU. It looks like it's more like 20% slower on your machine. I did make a note of that in the readme: "➖ Slower than GNU Core Utilities". Please let me know if you have any suggested changes here.